core/src/main/scala/kafka/log/ProducerStateManager.scala

```diff
@@ -107,13 +107,18 @@ private[log] class ProducerAppendInfo(val producerId: Long,
         throw new DuplicateSequenceNumberException(s"Duplicate sequence number for producerId $producerId: (incomingBatch.firstSeq, " +
           s"incomingBatch.lastSeq): ($firstSeq, $lastSeq), (lastEntry.firstSeq, lastEntry.lastSeq): " +
           s"(${this.firstSeq}, ${this.lastSeq}).")
-      } else if (firstSeq != this.lastSeq + 1L) {
+      } else if (!(firstSeq == this.lastSeq + 1L || (firstSeq == 0 && this.lastSeq == Short.MaxValue))) {
```

Comment:
I think this would be a little clearer if we had a `inSequence` method and then did the negation on it.